### PR TITLE
Java instanceof: avoid dereferencing null pointer

### PR DESCRIPTION
**Pull request by smowton**
Previously our code was of the form
```
tmp = x->@class_identifier
is_instanceof = x != null && tmp == "A" || tmp == "B" || ...
```

This was harmless as the value read from a null pointer was never used if it was null, but
would present a spurious dereference of a possibly-null pointer, introducing false uncertainty
about the vallue of `tmp`. Therefore we now generate:

```
tmp = x != null ? x->@class_identifier : ""
is_instanceof = x != null && tmp == "A" || tmp == "B" || ...
```

There should be no functional change regarding existing code.

### CODE REVIEW

**Review by thk123 — COMMENTED** [review]
LGTM but a TG pointer bump I think is a must for this.

File: jbmc/src/java_bytecode/remove_instanceof.cpp [hunk 1]
```diff
@@ -122,8 +131,6 @@ std::size_t remove_instanceoft::lower_instanceof(
   // disjunction of all possible object types. According to the Java
   // specification, null instanceof T is false for all possible values of T.
   // (http://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.20.2)
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/remove_instanceof.cpp:133]: maybe makes sense to adapt this comment now

**Comment by smowton** [comment]
https://github.com/diffblue/test-gen/pull/1932

File: jbmc/src/java_bytecode/remove_instanceof.cpp (hunk 1 above)
Comment (smowton) [reply jbmc/src/java_bytecode/remove_instanceof.cpp:133]: Disagree-- the code now generated has the form
```
clsid = ptr == null ? "" : ptr->@class_identifier
isA = ptr != null && clsid == "A"
```

The comment applies to the second line, and is placed prior to the second line as before. This change simply avoids deref'ing a pointer that can still be null.

Comment (mgudemann) [reply jbmc/src/java_bytecode/remove_instanceof.cpp:133]: ok then

**Review by mgudemann — APPROVED** [review]

**Comment by smowton** [comment]
Hard to tell, but it seems like this might cause massive slowdown on test-gen (that, or Travis is being shit today). This might be caused because instanceof tests are less likely to resolve by symex if it can't evaluate the `== null` test and so thinks the clsid is a superposition of a constant and `""`.

I'll have another go at this...

**Pull request closed without merging**